java-checks/src/main/java/org/sonar/java/checks/PublicStaticMutableMembersCheck.java

```diff
@@ -0,0 +1,155 @@
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections4.set.UnmodifiableSet")).name(DECORATE).withNoParameterConstraint())
+    .add(MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.apache.commons.collections4.list.UnmodifiableList")).name(DECORATE).withNoParameterConstraint());
+  private static final MethodMatcher ARRAYS_AS_LIST = MethodMatcher.create()
+    .typeDefinition("java.util.Arrays").name("asList").withNoParameterConstraint();
+
+  private static final Set<String> ACCEPTED_TYPES = ImmutableSet.of(
+    "com.google.common.collect.ImmutableMap",
+    "com.google.common.collect.ImmutableCollection"
+  );
+  
+  private static final Set<String> ACCEPTED_NEW_TYPES = ImmutableSet.of(
+    "org.apache.commons.collections4.list.UnmodifiableList"
+  );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.ENUM);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    List<Tree> members = ((ClassTree) tree).members();
+    for (Tree member : members) {
+      if (member.is(Tree.Kind.VARIABLE)) {
+        checkVariable((VariableTree) member);
+      }
+    }
+  }
+
+  private void checkVariable(VariableTree variableTree) {
+    Symbol symbol = variableTree.symbol();
+    if (symbol != null && isPublicStatic(symbol) && isForbiddenType(symbol.type()) && isNotImmutable(variableTree.initializer())) {
+      addIssue(variableTree, "Make this member \"protected\".");
+    }
+  }
+
+  static boolean isNotImmutable(@Nullable ExpressionTree initializer) {
+    if (initializer == null) {
+      return false;
+    }
+    ExpressionTree expression = ExpressionsHelper.skipParentheses(initializer);
+    if (expression.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) expression;
+      if (isAcceptedTypeOrUnmodifiableMethodCall(mit)) {
+        return false;
+      } else if (ARRAYS_AS_LIST.matches(mit)) {
+        return !mit.arguments().isEmpty();
+      }
+    } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+      return !isAcceptedType(expression.symbolType(), ACCEPTED_NEW_TYPES);
+    }
+    return true;
+  }
+
+  private static boolean isAcceptedTypeOrUnmodifiableMethodCall(MethodInvocationTree mit) {
+    return isAcceptedType(mit.symbolType(), ACCEPTED_TYPES) || UNMODIFIABLE_METHOD_CALLS.anyMatch(mit);
+  }
+
+  private static boolean isAcceptedType(Type type, Set<String> accepted) {
+    for (String acceptedType : accepted) {
+      if (type.isSubtypeOf(acceptedType)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static boolean isPublicStatic(Symbol symbol) {
+    return symbol.isStatic() && symbol.isPublic();
+  }
+
+  static boolean isForbiddenType(final Type type) {
+    return type.isArray() || Iterables.any(MUTABLE_TYPES, new Predicate<String>() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a named "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2694)